Fix start date calculation

katzmo opened the pull request.

katzmo committed:
Fix start date for single digit (day of) months.

katzmo committed:
Refine start date calculations.

katzmo committed:
Setup tests on github.

torotil commented on src/Utils.php:
```diff
@@ -137,28 +137,46 @@ public static function subscriptionData(\PaymentLineItem $line_item) {
+        $unit = 'year';
+      }
+      elseif ($interval_unit == 'monthly' && !empty($interval_value = $recurrence->interval_value)) {
+        $count = 12 % $interval_value == 0 ? $interval_value : $count;
```
It would be better to explicitly have `1` as else-value. We don’t want to keep the value (whatever it is) we want to set it to `1`.

katzmo replied:
Re-wrote the whole if statement, there was another glitch in it too (weekly payments without month). I hope it’s clearer now!

torotil reviewed (commented):
Perfect except for the comment and that’s debatable.

katzmo commented:
We’re clearly missing a _day of the week_ field for recurring payments, this would make the start date calculation even more interesting! :stuck_out_tongue_closed_eyes:

torotil commented on src/Utils.php:
```diff
@@ -137,28 +137,49 @@ public static function subscriptionData(\PaymentLineItem $line_item) {
+          $unit = 'year';
+        }
+        break;
+    }
```
I think it’s still a bit tricky to intepret. How about:

```suggestion
    $find_increment = function ($recurrence) {
      switch ($recurrence->interval_unit) {
        case 'weekly':
          return '1 day';
        case 'monthly':
          if (!empty($recurrence->month) && !empty($value = $recurrence->interval_value) && 12 % $value == 0) {
            return "$value month";
          }
          return '1 month';
        case 'yearly':
          return !empty($recurrence->month) ? '1 year' : '1 month';
      }
    };
    $increment = $find_increment($recurrence); 
```

… or even put this into a protected function. (Sorry I don’t know how to make suggestions for multiple lines at once.

This way each branch of the switch can be read on its own without contextual knowledge of the variables’ prior values.

katzmo replied:
You’re right, that’s more explicit, thanks!

torotil committed:
Deal with negative day_of_month values.

torotil committed:
Only allow start dates in 31-day months for negative day_of_month.

torotil committed:
Make start dates in the last days of the month work.

To make this easier the best way is perhaps to search forward and
backward from the modified start date.

torotil committed:
Simplify algorithm by using only 1 day increments.

Yes, this decreases performance slightly for some yearly payments.
Deciding for simpler code anyway as long is the performance hit is not
proven to be relevant.

torotil committed:
Simplify earliest date calculation using max().

torotil committed:
Add test for subscription generation.

torotil committed:
Remove 31-day-in-month-constraint for last day of month payments.

The resulting date might be too far in the future.

torotil committed:
Explicitly limit date iteration.

Stripe doesn’t allow start dates that are too much in the future. What’s
too much is not explicitly documented, but according to our tests the
start date has to be within one payment cycle from now.

torotil committed:
Ensure we have immutable dates when checking constraints.

torotil committed:
Fallback to last day of month if the actual day_of_month is too big.

torotil committed:
Fix undefined variable error.
